### ISSUE TITLE
Polymer 2.x typescript declaration files

### ISSUE THREAD

**Issue by gertcuykens** [Polymer/docs#1768]
Summary: 

Need to document Typescript type definitions, project set up and tool support for 2.x.
Depends on https://github.com/Polymer/polymer/issues/4025.

Original issue:

The polymer team is having a private 2.0 typescript party right now playing with their brand new 2.0 classes. Yes the 2.0 readme explains us how to build a 2.0 class very well, but it assumes setting up vscode is a walk in the park. This is what I have so far 

https://github.com/gertcuykens/hello-world
- type definitions not working for Polymer, customElements, import and shadowdom.
- bower is making a mess of things.
- can we still use iron-component-page somehow?
- not sure if the polymer 2 decorators are working?

Feel free to look at my example and document the complete setup so we can join the 2.0 typescript party. Thanks

![image](https://cloud.githubusercontent.com/assets/863731/18982152/38f6f140-86e4-11e6-8f06-31d2ff027096.png)

**Comment by robdodson** [comment]
@rictic @justinfagnani

**Comment by gertcuykens** [comment]
@robdodson Yesterday they where on slack polymer typescript group explaining the status. We just have to wait until somebody has time to finish the typescript definitions first before we can move forward. 

https://github.com/Polymer/polymer/issues/4025

You can try out https://github.com/gertcuykens/hello-world to see the error messages and compatibility issues.

**Comment by gertcuykens** [comment]
@nippur72 Callout to the PolymerTS developer if he could give a hand for typings for Polymer 2 :D

**Comment by rictic** [comment]
I'm working on typings for shadow dom and custom elements.

We don't have any private Typescript code that we're not sharing, it's just a work in progress :)

**Comment by rictic** [comment]
https://github.com/Microsoft/TSJS-lib-generator/pull/150 and https://github.com/Microsoft/TSJS-lib-generator/pull/151

**Comment by gertcuykens** [comment]
Yeaaaa, thanks, things are going to get great in polymer typescript world :D 

PS What about https://w3c.github.io/webcomponents/spec/imports/ typings ? :)

**Comment by rictic** [comment]
Already in, as it turns out! You just need to cast the element to an HTMLLinkElement. I think something like this would work:

```typescript
const l: HTMLLinkElement = document.querySelector('link[rel="import"]');
```

**Comment by gertcuykens** [comment]
Hmm can't figure it out

```js
error TS2322: Type 'Element | null' is not assignable to type 'HTMLLinkElement'.                                                                                           
                         Type 'null' is not assignable to type 'HTMLLinkElement'. 
```

![image](https://cloud.githubusercontent.com/assets/863731/19102723/bb31df84-8ad2-11e6-9294-9c44bb9444c5.png)

PS I made some changes to https://github.com/gertcuykens/hello-world i have less tsc errors but  still allot of things that can't be right :)

**Comment by rictic** [comment]
Maybe try this?

```typescript
const l = document.querySelector('link[rel="import"]') as  HTMLLinkElement;
```

**Comment by gertcuykens** [comment]
Yep working thanks :) 4 more errors to go and figuring out why the decorator stuff is doing notthing

![image](https://cloud.githubusercontent.com/assets/863731/19127483/7b1f3d84-8b3f-11e6-9ef2-d70df5845300.png)

I would like to pay your attention on `connectedCallback` etc because I think It has to do with the polymer typings specifically?
```js
    connectedCallback() {
        super.connectedCallback();
    }
```
![image](https://cloud.githubusercontent.com/assets/863731/19127520/9bcf7e68-8b3f-11e6-93c0-595d982e928b.png)

**Comment by rictic** [comment]
Sent a PR for two of the errors: https://github.com/gertcuykens/hello-world/pull/1

The `connectedCallback()` issue looks like it's correct. `Polymer.Element` doesn't have a defined `connectedCallback` afaict.

**Comment by gertcuykens** [comment]
Hmm but HTMLElement does right?
So I can conclude Polymer.Element must inherit all does native callbacks too right?

https://github.com/Polymer/polymer/tree/2.0-preview#20-es6-class-based-syntax

**Comment by rictic** [comment]
I don't think HTMLElement provides a default implementation of connectedCallback:

<img width="230" alt="screen shot 2016-10-05 at 3 20 48 pm" src="https://cloud.githubusercontent.com/assets/1659/19133855/691ceb52-8b0f-11e6-942a-df0d9c2d99ad.png">

**Comment by gertcuykens** [comment]
Ok, thanks, so that means the example code in the 2.0 review readme is wrong or do I need to do something different in my typescript?

https://github.com/Polymer/polymer/tree/2.0-preview#20-es6-class-based-syntax

**Comment by rictic** [comment]
I don't actually know. Someone should try it out with native ES6 in the polymer 2.0-preview branch. If there's a runtime error then we should fix the readme, otherwise we should fix the typings for `Polymer.Element`

**Comment by gertcuykens** [comment]
As far as I can tell I have no runtime errors in my browser running my example, so I think its going to be the typings that are incomplete?

![image](https://cloud.githubusercontent.com/assets/863731/19135034/12699b40-8b62-11e6-871e-6279de261df0.png)

**Comment by arthurevans** [comment]
`connectedCallback` is definitely on there:

https://github.com/Polymer/polymer/blob/2.0-preview/src/elements/element.html#L219

That's where all the property effects get kicked off. Without `connectedCallback`, all would be darkness and chaos.

**Comment by gertcuykens** [comment]
Thanks @arthurevans , can you go through the list of properties and methods please with @rictic and @justinfagnani before a incomplete list is submitted to microsoft or merged into 2.0 preview

**Comment by gertcuykens** [comment]
Nooo my life is getting complicated :( It only has a runtime error on a HTMLElement not a Polymer.Element so that means a Polymer Element is not a HTMLElement. This is bad :(

```js
class HelloWorldES extends HTMLElement {
  _hello: string = "Hello World"
  constructor() {
    super()
    const l = document.querySelector('link[rel="import"]') as HTMLLinkElement
    if (l.import == null) return;
    const t = l.import.querySelector('template')
    if (t == null) return;
    const instance = t.content.cloneNode(true)
    const shadowRoot = this.attachShadow({ mode: 'open' })
    shadowRoot.appendChild(instance)
  }
  connectedCallback() {
    super.connectedCallback()
    super.disconnectedCallback()
  }
  attributeChangedCallback(name: string, oldValue: string, newValue: string) { }
  adoptedCallback() { }
}
customElements.define('hello-world-es', HelloWorldES)
```
<img width="619" alt="screen shot 2016-10-06 at 21 48 14" src="https://cloud.githubusercontent.com/assets/863731/19168085/c859aede-8c0e-11e6-9ad3-2fd4e3329db9.png">

**Comment by rictic** [comment]
No, it just means that `HTMLElement` doesn't have a `connectedCallback` method but `Polymer.Element` does.

Custom elements _may_ define any of the lifecycle callback methods, but any given custom element may or may not. It sounds like Polymer.Element does define some, and our typings should reflect this.

**Comment by gertcuykens** [comment]
PS What would happen if we made `connectedCallback` and the others _private in Polymer.Element src? Would that break everything? 

https://github.com/Polymer/polymer/blob/2.0-preview/src/elements/element.html#L219-L226

**Comment by rictic** [comment]
I can see no reason to make them private, as extending classes must call them via super.connectedCallback(), and we can't prefix them with underscores, as the spec dictates how they are named.

**Comment by gertcuykens** [comment]
Ok thanks, I think I understand your point about PolymerElements extending each other. I am going to try to find out if it is actually possible to do the same with vanialla HTMLElements extending each other and how that would look like. And If `super.connectedCallback()` pops up in that senerio.

**Comment by gertcuykens** [comment]
Yeaaay its working without runtime errors :D I am getting the complete picture now

```js
class HelloWorldES extends HTMLElement {
  _hello: string = "Hello World"
  constructor() {
    super()
    const l = document.querySelector('link[rel="import"]') as HTMLLinkElement
    if (l.import == null) return;
    const t = l.import.querySelector('template')
    if (t == null) return;
    const instance = t.content.cloneNode(true)
    const shadowRoot = this.attachShadow({ mode: 'open' })
    shadowRoot.appendChild(instance)
  }
  connectedCallback() { }
  disconnectedCallback() { }
  attributeChangedCallback(name: string, oldValue: string, newValue: string) { }
  adoptedCallback() { }
}

class HelloWorldES2 extends HelloWorldES {
  constructor() {
    super()
  }
  connectedCallback() {
    super.connectedCallback()
  }
}
customElements.define('hello-world-es', HelloWorldES2)
```

**Comment by katejeffreys** [comment]
If this should still be an open issue for the Polymer docs, could someone outline the documentation requirements?

@arthurevans @rictic @gertcuykens

**Comment by arthurevans** [comment]
We can't do anything about this in docs until eng finalizes Typescript declarations and figures out how/where to publish them (for example, as part of Polymer or in a separate repo).

**Comment by rictic** [comment]
I think it's fair to assign this to Justin, as he's taking the lead on this last we checked.

**Comment by arthurevans** [comment]
That's fine. There's an issue (linked in the summary at the top, now) for TS support on Polymer/polymer. When that's implemented and we have eng docs, Justin can assign this back to one of us.

**Comment by arthurevans** [comment]
Closing this as it's not tracking anything useful. We'll document TS support when we have it.
